File: src/main/scala/imcode/server/document/index/service/impl/DocumentIndexServiceOps.scala
package imcode.server.document.index.service.impl

import com.imcode._
import com.imcode.imcms.mapping.DocumentMapper

import _root_.imcode.server.document.DocumentDomainObject
import _root_.imcode.server.document.index.DocumentIndex
import _root_.imcode.server.document.index.service.IndexRebuildProgress

import com.imcode.imcms.api.DocumentLanguage
import scala.collection.SeqView
import scala.collection.JavaConverters._

import java.util.Date
import java.net.URLDecoder

import org.apache.solr.common.SolrInputDocument
import org.apache.solr.common.util.DateUtil
import org.apache.solr.client.solrj.{SolrQuery, SolrServer}
import org.apache.solr.client.solrj.response.QueryResponse


/**
 * Document index service low level operations.
 *
 * An instance of this class is thread save.
 */
// todo: document search might return doc which is not present in db (deleted) - return stub instead
class DocumentIndexServiceOps(documentMapper: DocumentMapper, documentIndexer: DocumentIndexer) extends Log4jLoggerSupport {

  type DocId = Int


  @throws(classOf[SolrInputDocumentCreateException])
  private def withExceptionWrapper[A](body: => A): A = {
    try {
      body
    } catch {
      case e: Exception => throw new SolrInputDocumentCreateException(e)
    }
  }


  @throws(classOf[SolrInputDocumentCreateException])
  def mkSolrInputDocs(docId: Int): Seq[SolrInputDocument] = withExceptionWrapper {
    mkSolrInputDocs(docId, documentMapper.getImcmsServices.getDocumentLanguageSupport.getLanguages.asScala)
  }


  @throws(classOf[SolrInputDocumentCreateException])
  def mkSolrInputDocs(docId: Int, languages: Seq[DocumentLanguage]): Seq[SolrInputDocument] = withExceptionWrapper {
    val solrInputDocs = for {
      language <- languages
      doc <- documentMapper.getDefaultDocument[DocumentDomainObject](docId, language).asOption
    } yield documentIndexer.index(doc)


    if (logger.isTraceEnabled) {
      logger.trace("Created %d solrInputDoc(s) with docId: %d and language(s): %s.".format(
        solrInputDocs.length, docId, languages.mkString(", "))
      )
    }

    solrInputDocs
  }


  @throws(classOf[SolrInputDocumentCreateException])
  def mkSolrInputDocsView(): SeqView[(DocId, Seq[SolrInputDocument]), Seq[_]] = {
    documentMapper.getImcmsServices.getDocumentLanguageSupport.getLanguages.asScala |> {
      languages =>
        documentMapper.getAllDocumentIds.asScala.view.map(docId => docId.toInt -> mkSolrInputDocs(docId, languages))
    }
  }


  def mkSolrDocsDeleteQuery(docId: Int): String = "%s:%d".format(DocumentIndex.FIELD__META_ID, docId)


  def query(solrServer: SolrServer, solrQuery: SolrQuery): QueryResponse = {
    if (logger.isDebugEnabled) {
      logger.debug("Searching using SOLr query: %s.".format(URLDecoder.decode(solrQuery.toString, "UTF-8")))
    }

    solrServer.query(solrQuery)
  }


  @throws(classOf[SolrInputDocumentCreateException])
  def addDocsToIndex(solrServer: SolrServer, docId: Int) {
    mkSolrInputDocs(docId) |> {
      solrInputDocs =>
        if (solrInputDocs.nonEmpty) {
          solrServer.add(solrInputDocs.asJava)
          solrServer.commit()
          logger.trace("added %d solrInputDoc(s) with docId %d into the index.".format(solrInputDocs.length, docId))
        }
    }
  }


  def deleteDocsFromIndex(solrServer: SolrServer, docId: Int): Unit = mkSolrDocsDeleteQuery(docId) |> {
    deleteQuery =>
      solrServer.deleteByQuery(deleteQuery)
      solrServer.commit()
  }


  @throws(classOf[InterruptedException])
  @throws(classOf[SolrInputDocumentCreateException])
  def rebuildIndex(solrServer: SolrServer)(progressCallback: IndexRebuildProgress => Unit) {
    logger.debug("Rebuilding index.")
    val docsView = mkSolrInputDocsView()
    val docsCount = docsView.length
    val rebuildStartDt = new Date
    val rebuildStartMills = rebuildStartDt.getTime

    progressCallback(IndexRebuildProgress(rebuildStartMills, rebuildStartMills, docsCount, 0))

    for {
      ((docId, solrInputDocs), docNo) <- docsView.zip(Stream.from(1)); if solrInputDocs.nonEmpty
    } {
      if (Thread.interrupted()) {
        solrServer.rollback()
        throw new InterruptedException()
      }

      solrServer.add(solrInputDocs.asJava)
      logger.debug(s"Added input docs [$solrInputDocs] to index.")
      progressCallback(IndexRebuildProgress(rebuildStartMills, new Date().getTime, docsCount, docNo))
    }

    logger.debug("Deleting old documents from index.")
    solrServer.deleteByQuery("timestamp:{* TO %s}".format(DateUtil.getThreadLocalDateFormat.format(rebuildStartDt)))
    solrServer.commit()
    logger.debug("Index rebuild is complete.")
  }
}

File: src/main/java/imcode/server/document/index/service/impl/DocumentIndexServiceOps.java
package imcode.server.document.index.service.impl;

import com.imcode.imcms.mapping.DocumentMapper;
import imcode.server.document.index.DocumentIndex;
import lombok.SneakyThrows;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.solr.client.solrj.SolrClient;
import org.apache.solr.client.solrj.SolrQuery;
import org.apache.solr.client.solrj.SolrServerException;
import org.apache.solr.client.solrj.response.QueryResponse;
import org.apache.solr.common.SolrInputDocument;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.Date;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Consumer;

/**
 * Document index service low level operations.
 * <p>
 * An instance of this class is thread save.
 */
@Component
public class DocumentIndexServiceOps {

    private static final Logger logger = LogManager.getLogger(DocumentIndexServiceOps.class);

    private final DocumentMapper documentMapper;
    private final DocumentIndexer documentIndexer;

    private AtomicLong indexedDocumentsAmount = new AtomicLong(-1);

    @Autowired
    public DocumentIndexServiceOps(DocumentMapper documentMapper,
                                   DocumentIndexer documentIndexer) {

        this.documentMapper = documentMapper;
        this.documentIndexer = documentIndexer;
    }

    private SolrInputDocument mkSolrInputDoc(int docId) {
        try {
            return documentIndexer.index(docId);
        } catch (Exception e) {
            logger.error(
                    String.format("Can't create SolrInputDocument from doc %d", docId),
                    e);
            return null;
        }
    }

	private SolrInputDocument prepareSolrInputDocForVersionUpdate(int docId){
		try {
			return documentIndexer.updateDocumentVersion(docId);
		}catch (Exception e){
			logger.error(
					String.format("Can`t prepare SolrInputDocument for update from doc %d", docId), e
			);
			return null;
		}
	}

    private String mkSolrDocsDeleteQuery(int docId) {
        return String.format("%s:%d", DocumentIndex.FIELD__META_ID, docId);
    }

    public QueryResponse query(SolrClient solrClient, SolrQuery solrQuery) throws SolrServerException, IOException {
        return solrClient.query(solrQuery);
    }

    public void addDocsToIndex(SolrClient solrClient, int docId) throws SolrServerException, IOException {
        final SolrInputDocument solrInputDoc = mkSolrInputDoc(docId);

        if (solrInputDoc != null) {
            solrClient.add(solrInputDoc);
            solrClient.commit(false, false, true);

            logger.error(String.format("Added solrInputDoc with docId %d into the index.", docId));
        }
    }

    public void deleteDocsFromIndex(SolrClient solrClient, int docId) throws SolrServerException, IOException {
        String query = mkSolrDocsDeleteQuery(docId);

        solrClient.deleteByQuery(query);
        solrClient.commit(false, false, true);
        logger.info(String.format("Removed document with docId %d from index.", docId));
    }

	public void updateDocumentVersionInIndex(SolrClient solrClient, int docId) throws SolrServerException, IOException {
		final SolrInputDocument solrInputDocument = prepareSolrInputDocForVersionUpdate(docId);

		if (solrInputDocument != null) {
			solrClient.add(solrInputDocument);
			solrClient.commit(false, false, true);

			logger.error(String.format("Updated document version in index with docId %d.", docId));
		}
	}

    public void rebuildIndex(SolrClient solrClient) {
        rebuildIndex(solrClient, indexRebuildProgress -> {
        });
    }

    @SneakyThrows
    private void rebuildIndex(SolrClient solrClient, Consumer<IndexRebuildProgress> progressCallback) {
        logger.debug("Rebuilding index.");

        final List<Integer> ids = documentMapper.getAllDocumentIds();

        final int docsCount = ids.size();
        int docNo = 0;
        final Date rebuildStartDt = new Date();
        final long rebuildStartTime = rebuildStartDt.getTime();

        progressCallback.accept(new IndexRebuildProgress(rebuildStartTime, rebuildStartTime, docsCount, docNo));

        indexedDocumentsAmount.set(0);

        for (int id : ids) {
            if (Thread.interrupted()) {
                solrClient.rollback();
                throw new InterruptedException();
            }

            SolrInputDocument solrInputDoc = mkSolrInputDoc(id);
            if (solrInputDoc != null) {
                solrClient.add(solrInputDoc);
                logger.debug("Added input doc with id " + id + " to index.");
            }
            indexedDocumentsAmount.incrementAndGet();

            docNo += 1;
            progressCallback.accept(new IndexRebuildProgress(rebuildStartTime, System.currentTimeMillis(), docsCount, docNo));
        }

        logger.debug("Deleting old documents from index.");

        solrClient.deleteByQuery(String.format("timestamp:{* TO %s}", rebuildStartDt.toInstant().toString()));
        solrClient.commit();
        indexedDocumentsAmount.set(-1);

        logger.debug("Index rebuild is complete.");
    }

    public long getAmountOfIndexedDocuments() {
        return indexedDocumentsAmount.get();
    }
}
